src/com/amazon/ion/impl/ImportLocation.java

```diff
@@ -0,0 +1,53 @@
+package com.amazon.ion.impl;
+
+/**
+ * A SymbolToken's import location, allowing for symbols with unknown text to be mapped to a particular slot
+ * in a shared symbol table.
+ * NOTE: this is currently not publicly accessible, but it is an important step toward being able to correctly
+ * round-trip symbols with unknown text from shared symbol tables in different symbol table contexts. See
+ * https://github.com/amzn/ion-java/issues/126 . Support is added now to avoid risking the appearance of performance
+ * degradation if ImportLocation support were added after initial release of the incremental IonReader.
+ */
+class ImportLocation {
+
+    // The name of the shared symbol table.
+    final String name;
+
+    // The index into the shared symbol table.
+    final int sid;
+
+    ImportLocation(String name, int sid) {
+        this.name = name;
+        this.sid = sid;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public int getSid() {
+        return sid;
+    }
+
+    @Override
+    public String toString() {
+        return String.format("ImportLocation::{name: %s, sid: %d}", name, sid);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (!(o instanceof ImportLocation)) {
+            return false;
+        }
+        ImportLocation that = (ImportLocation) o;
+        return this.getName().equals(that.getName()) && this.getSid() == that.getSid();
+    }
+
+    @Override
+    public int hashCode() {
+        int result = 17;
+        result += 31 * getName().hashCode();
+        result += 31 * getSid();
+        return result;
+    }
+}
```

Comment:
I think the biggest hassle with `Arrays.hashcode(Object[])` is its signature. It's not `Arrays.hashCode(Object...)`, so we can't actually call it like `Arrays.hashCode(getName(), getSid())`...

What we want is something like `Objects.hash(Object...)`. But alas, `Objects` was added in JDK 1.7...

Reply:
The implementation of `Objects.hash` is:
```java
    public static int hash(Object... values) {
        return Arrays.hashCode(values);
    }
```

Varargs are really arrays, the `...` is just syntactic sugar. `public static void main(String... args)` is the exact same thing as the canonical `String[] args`. 

As an aside, it's amusing that even the Java syntax highlighter that GitHub uses apparently doesn't know what to do with `...`.

Reply:
By the "state vector" idiom I mean something like this (less interesting parts of class elided):
```java
        ParsedMessage(Object message) {
            Matcher matcher = MESSAGE_PATTERN.matcher(message.toString());
            if (!matcher.matches()) {
                throw new IllegalArgumentException(String.format("cannot parse '%s'", message));
            }
            identifier = matcher.group(1);
            tag = matcher.group(2);
            stackHash = matcher.group(3);

            state = new Object[]{identifier, tag, stackHash};
        }

        @Override
        public boolean equals(Object o) {
            if (!(o instanceof ParsedMessage)) {
                return false;
            }
            ParsedMessage that = (ParsedMessage) o;
            return Arrays.equals(this.state, that.state);
        }

        @Override
        public int hashCode() {
            return Arrays.hashCode(state);
        }
```

Then you only need update the `state` in order to maintain `equals` and `hashCode`. On the other hand, something like Immutables can [generate](https://immutables.github.io/generated.html) the same kind of optimal hashcode generation (no autoboxing or iterator-style iteration) as you have above with even less maintenance burden so that's an even better way to go :D